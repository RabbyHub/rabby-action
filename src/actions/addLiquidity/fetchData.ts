import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
