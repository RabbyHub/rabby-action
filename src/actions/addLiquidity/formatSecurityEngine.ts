import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
