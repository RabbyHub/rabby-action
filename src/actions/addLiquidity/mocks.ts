export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
