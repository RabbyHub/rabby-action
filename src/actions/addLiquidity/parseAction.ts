import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
