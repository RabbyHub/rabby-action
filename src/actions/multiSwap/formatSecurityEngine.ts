import {
  AssetOrderRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineMultiSwap: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.multiSwap) {
      return {};
    }
    const { receiver } = actionData.multiSwap;
    const data = requireData as AssetOrderRequireData;
    return {
      multiSwap: {
        from: data.sender,
        receiver: receiver || '',
        chainId,
        id: data.id,
      },
    };
  };
