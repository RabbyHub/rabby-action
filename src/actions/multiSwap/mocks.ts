export const preExecData = {
  abi: {
    func: 'swapMultiCompact',
    params: [],
  },
  abi_str: '{\n    "func": "swapMultiCompact",\n    "params": []\n}',
  balance_change: {
    success: true,
    error: null,
    send_token_list: [
      {
        id: '0x50c5725949a6f0c72e6c4a641f24049a917db0cb',
        chain: 'base',
        name: 'Dai Stablecoin',
        symbol: 'DAI',
        display_symbol: null,
        optimized_symbol: 'DAI',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/eth_token/logo_url/0x6b175474e89094c44da98b954eedeac495271d0f/549c4205dbb199f1b8b03af783f35e71.png',
        protocol_id: '',
        price: 0.9998,
        price_24h_change: 0.0,
        credit_score: 1653203.344730977,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1689928543.0,
        amount: 102.05317125734108,
        usd_value: 102.03276062308962,
        raw_amount: '102053171257341079245',
      },
      {
        id: '0xd9aaec86b65d86f6a7b5b1b0c42ffa531710b6ca',
        chain: 'base',
        name: 'USD Base Coin',
        symbol: 'USDbC',
        display_symbol: null,
        optimized_symbol: 'USDbC',
        decimals: 6,
        logo_url:
          'https://static.debank.com/image/base_token/logo_url/0xd9aaec86b65d86f6a7b5b1b0c42ffa531710b6ca/02eafdd0ee21a4de9f19a6bc861f3024.png',
        protocol_id: 'base_fxdx',
        price: 0.9999000099990001,
        price_24h_change: 9.999000099986184e-5,
        credit_score: 5640170.856598627,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1690914161.0,
        amount: 0.997474,
        usd_value: 0.9973742625737426,
        raw_amount: '997474',
      },
    ],
    receive_token_list: [
      {
        id: 'base',
        chain: 'base',
        name: 'ETH',
        symbol: 'ETH',
        display_symbol: null,
        optimized_symbol: 'ETH',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
        protocol_id: '',
        price: 2687.31,
        price_24h_change: -0.016855700278062238,
        credit_score: 100000000.0,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: null,
        amount: 0.03831174429709561,
        usd_value: 102.955533567028,
        raw_amount: '38311744297095608',
      },
    ],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: -0.07460131863535935,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 449340,
    gas_limit: 511491,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 2687.31,
    price_24h_change: -0.016855700278062238,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: null,
    amount: 0.04635428024034957,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_call: {
    action: 'swapMultiCompact',
    contract: '0x19ceead7105607cd444f5ad10dd51356436095a1',
    contract_protocol_name: 'ODOS',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/arb_odos/c5e434b52cd261ee06b2ff75f4b5c1ea.png',
  },
  trace_id: 'ffd69513ac1c4c11ba56138e3c4090d0',
} as any;

export const parseTxData = {
  action: {
    type: 'multi_swap_token',
    data: {
      pay_tokens: [
        {
          id: '0x50c5725949a6f0c72e6c4a641f24049a917db0cb',
          chain: 'base',
          name: 'Dai Stablecoin',
          symbol: 'DAI',
          display_symbol: null,
          optimized_symbol: 'DAI',
          decimals: 18,
          logo_url:
            'https://static.debank.com/image/eth_token/logo_url/0x6b175474e89094c44da98b954eedeac495271d0f/549c4205dbb199f1b8b03af783f35e71.png',
          protocol_id: '',
          price: 0.9998,
          price_24h_change: 0.0,
          credit_score: 1653203.344730977,
          is_verified: true,
          is_scam: false,
          is_suspicious: false,
          is_core: true,
          is_wallet: true,
          time_at: 1689928543.0,
          amount: 102.05317125734108,
          raw_amount: '102053171257341079245',
        },
        {
          id: '0xd9aaec86b65d86f6a7b5b1b0c42ffa531710b6ca',
          chain: 'base',
          name: 'USD Base Coin',
          symbol: 'USDbC',
          display_symbol: null,
          optimized_symbol: 'USDbC',
          decimals: 6,
          logo_url:
            'https://static.debank.com/image/base_token/logo_url/0xd9aaec86b65d86f6a7b5b1b0c42ffa531710b6ca/02eafdd0ee21a4de9f19a6bc861f3024.png',
          protocol_id: 'base_fxdx',
          price: 0.9999000099990001,
          price_24h_change: 9.999000099986184e-5,
          credit_score: 5640170.856598627,
          is_verified: true,
          is_scam: false,
          is_suspicious: false,
          is_core: true,
          is_wallet: true,
          time_at: 1690914161.0,
          amount: 0.997474,
          raw_amount: '997474',
        },
      ],
      receive_tokens: [
        {
          id: 'base',
          chain: 'base',
          name: 'ETH',
          symbol: 'ETH',
          display_symbol: null,
          optimized_symbol: 'ETH',
          decimals: 18,
          logo_url:
            'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
          protocol_id: '',
          price: 2687.31,
          price_24h_change: -0.016855700278062238,
          credit_score: 100000000.0,
          is_verified: true,
          is_scam: false,
          is_suspicious: false,
          is_core: true,
          is_wallet: true,
          time_at: null,
          min_amount: 0.03814282126978977,
          min_raw_amount: '3.814282126978977e+16',
        },
      ],
      receiver: '0x341a1fbd51825e5a107db54ccb3166deba145479',
    },
  },
  contract_call: {
    func: 'swapMultiCompact',
    contract: {
      id: '0x19ceead7105607cd444f5ad10dd51356436095a1',
      protocol: {
        name: 'ODOS',
        logo_url:
          'https://static.debank.com/image/project/logo_url/arb_odos/c5e434b52cd261ee06b2ff75f4b5c1ea.png',
      },
    },
  },
  log_id: 89081565,
} as any;

export const txData = {
  chainId: 8453,
  data: '0x84a7f3dd02010001f73F77F9466Da712590ae432A80f07fd50A7de600b087b5b5851380500000000000509058845b28ec3eb6ecd00015Fb33b095c6e739BE19364Ab408CD8f102262bB60003030f38620000000004100606020001341a1fBD51825E5a107DB54cCb3166DeBA14547900000001080303080206020202030c659734f1eef9c63b7ebdf78a164cdd745586db000000000000000000000046003401000004050180000006040300000001030019020a0000060700000401000201ff00000000000000000000000000000000000000009a0b05f3cf748a114a4f8351802b3bffe07100d4d9aaec86b65d86f6a7b5b1b0c42ffa531710b6caeb466342c4d449bc9f53a865d5cb90586f4052155fb33b095c6e739be19364ab408cd8f102262bb650c5725949a6f0c72e6c4a641f24049a917db0cb74cb6260be6f31965c239df6d6ef2ac2b5d4f020833589fcd6edb6e08f4c7c32d4f71b54bda029130000000000000000000000000000000000000000',
  from: '0x341a1fbd51825e5a107db54ccb3166deba145479',
  gas: '0x0',
  maxFeePerGas: '0x50df20',
  maxPriorityFeePerGas: '0x50df20',
  nonce: '0x69',
  to: '0x19ceead7105607cd444f5ad10dd51356436095a1',
  value: '0x0',
};
