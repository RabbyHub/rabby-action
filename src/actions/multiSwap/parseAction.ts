import { MultiSwapAction } from '@rabby-wallet/rabby-api/dist/types';
import { ParseAction } from '../../types';

export const parseActionMultiSwap: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'multi_swap_token') {
    return {};
  }

  const {
    pay_token_list: payTokenList,
    receive_token_list: receiveTokenList,
    receiver,
  } = data.data as MultiSwapAction;

  return {
    multiSwap: {
      payTokenList,
      receiveTokenList,
      receiver,
    },
  };
};
