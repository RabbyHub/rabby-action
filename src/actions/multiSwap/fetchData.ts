import { FetchActionRequiredData } from '../../types';
import { fetchDataSwap } from '../swap/fetchData';

export const fetchDataMultiSwap: FetchActionRequiredData = async (options) => {
  if (!options.actionData.multiSwap) {
    return {};
  }
  return fetchDataSwap(options, options.actionData.multiSwap);
};
