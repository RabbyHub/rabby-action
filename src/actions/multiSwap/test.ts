import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { parseTxData, preExecData, txData } from './mocks';
import { parseActionMultiSwap } from './parseAction';
import { fetchDataMultiSwap } from './fetchData';
import { formatSecurityEngineMultiSwap } from './formatSecurityEngine';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * https://app.odos.xyz/
 * multi from token
 */
test.each([
  [parseActionMultiSwap, fetchDataMultiSwap, formatSecurityEngineMultiSwap],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('MultiSwap', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionMultiSwap');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });
  expect(requireData).toMatchSnapshot('fetchDataMultiSwap');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineMultiSwap');
});
