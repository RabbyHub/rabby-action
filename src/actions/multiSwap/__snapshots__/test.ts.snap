// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSwap: fetchDataMultiSwap 1`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
  },
  "hasInteraction": true,
  "id": "0x19ceead7105607cd444f5ad10dd51356436095a1",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": true,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`MultiSwap: fetchDataMultiSwap 2`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
  },
  "hasInteraction": true,
  "id": "0x19ceead7105607cd444f5ad10dd51356436095a1",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": true,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`MultiSwap: formatSecurityEngineMultiSwap 1`] = `
{
  "multiSwap": {
    "chainId": "eth",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0x19ceead7105607cd444f5ad10dd51356436095a1",
    "receiver": "0x341a1fbd51825e5a107db54ccb3166deba145479",
  },
}
`;

exports[`MultiSwap: formatSecurityEngineMultiSwap 2`] = `
{
  "multiSwap": {
    "chainId": "eth",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0x19ceead7105607cd444f5ad10dd51356436095a1",
    "receiver": "0x341a1fbd51825e5a107db54ccb3166deba145479",
  },
}
`;

exports[`MultiSwap: parseActionMultiSwap 1`] = `
{
  "multiSwap": {
    "payTokenList": undefined,
    "receiveTokenList": undefined,
    "receiver": "0x341a1fbd51825e5a107db54ccb3166deba145479",
  },
}
`;

exports[`MultiSwap: parseActionMultiSwap 2`] = `
{
  "multiSwap": {
    "payTokenList": undefined,
    "receiveTokenList": undefined,
    "receiver": "0x341a1fbd51825e5a107db54ccb3166deba145479",
  },
}
`;
